jax/numpy/lax_numpy.py

```diff
@@ -2645,6 +2645,17 @@ def polyadd(a, b):
     return b.at[-a.shape[0]:].add(a)
 
 
+@_wraps(np.polyder)
+def polyder(a, m=1):
+  a = asarray(a)
+  if m < 0:
+      raise ValueError("Order of derivative must be positive")
+
+  if m == 0:
+    return a
+  return polyder(a[:-1] * arange(len(a) - 1, 0, -1), m - 1)
```

Comment:
Given the mechanics of JAX jit compilation, I think we should avoid this recursive approach. This function will have to be compiled with ``m`` as a static value, meaning that computing it via recursion would require `m` compilation steps. That said, a straightforward iterative replacement (based on for-loops) would also lead to slow compilation. It would be worth thinking about whether this could be expressed in terms of `jax.numpy` primitives: I suspect the answer is yes, with a bit of thought.

Reply:
I think this works:
```python
def polyder(a, m=1):
  a = asarray(a)
  if m < 0:
    raise ValueError("Order of derivative must be positive")
  if m % 1:
    raise ValueError("m must be an integer")
  N = len(a)
  coeff = (arange(len(a), m, -1) - 1 - arange(m)[:, newaxis]).prod(0)
  return a[:-m] * coeff
```

Reply:
Thanks for the insight! (Although I guess this makes jit pretty restrictive in terms of usage)
The above code gives me an error: `newaxis not defined`

Reply:
Yes, jit is pretty restrictive, but is quite useful in the contexts it's designed for (note the issue is not that the above code will not compile; the issue is that as written the one-time compilation cost grows with `m`.

`newaxis` is defined in the file here: https://github.com/google/jax/blob/05fde0f1709783610e3ccaaaf8ba682a9691c750/jax/numpy/lax_numpy.py#L62

Reply:
@jakevdp Thanks! Yes I did import these but didn't realize that `newaxis` is defined beyond function scope. Looking ahead, the function is giving me incorrect results. `polyder([1,1,1], m=1)` should give me `[2,1]` but I get `[2,2]`.

I'm not quite sure if `a[:-m] * coeff` is the right thing to do, given that each coefficient will need to be multiplied by a different factor (here we multiply each element in `a[:-m]` by a single number: coeff).
It might be a bit tricky to think of an approach without using loops and recursion here.

Reply:
I get `[2, 1]` when I run `polyder([1, 1, 1], m=1)` with the code above. Regarding the multiple coefficients: that is accomplished by the ``prod()`` reduction above. I fixed an issue with `m = 0`, and it seems to produce the expected result for a variety of inputs:
```
import numpy as np
from numpy.testing import assert_allclose
from jax.numpy import newaxis, arange, asarray

def polyder(a, m=1):
  a = asarray(a)
  if m < 0:
    raise ValueError("Order of derivative must be positive")
  if m == 0:
    return a
  if m % 1:
    raise ValueError("m must be an integer")
  coeff = (arange(len(a), m, -1) - 1 - arange(m)[:, newaxis]).prod(0)
  return a[:-m] * coeff

for nterms in range(10):
  for m in range(10):
    p = np.random.randint(-10, 10, nterms)
    assert_allclose(np.polyder(p, m), polyder(p, m))
```

Reply:
Ahh, my bad, I was reading `arange(m)[:, newaxis]` as `arange(m)[:newaxis]`. Apologies